packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -834,24 +798,37 @@ add (const Scalar& alpha,
     }
 #ifdef HAVE_TPETRA_MMM_TIMINGS
     MM = Teuchos::null;
-    MM = rcp(new TimeMonitor(*TimeMonitor::getNewTimer(prefix_mmm + std::string("building optimized column map"))));
+    MM = rcp(new TimeMonitor(*TimeMonitor::getNewTimer(prefix_mmm + std::string("Constructing graph"))));
 #endif
-    CrowMap = Bprime->getRowMap();
-    //OLD: CcolMap = AddKern::makeColMapAndConvertGids(Aprime->getGlobalNumCols(), globalColinds, colinds, comm);
-    //Get C's column map as the union of Aprime and Bprime col maps
-    CcolMap = AddKern::makeColMapAndConvertGids(Aprime->getGlobalNumCols(), globalColinds, colinds, comm);
+    RCP<const map_type> CcolMap;
+    Tpetra::Details::makeColMap<LocalOrdinal, GlobalOrdinal, Node>
+      (CcolMap, Aprime->getDomainMap(), globalColinds);
+    C.replaceColMap(CcolMap);
+    col_inds_array localColinds("C colinds", globalColinds.extent(0));
+    Kokkos::parallel_for(Kokkos::RangePolicy<exec_space>(0, globalColinds.extent(0)),
+        ConvertGlobalToLocalFunctor<LocalOrdinal, GlobalOrdinal,
+                              col_inds_array, global_col_inds_array,
+                              typename map_type::local_map_type>
+        (localColinds, globalColinds, CcolMap->getLocalMap()));
+    C.setAllValues(rowptrs, localColinds, vals);
+    //Use a regular fillComplete so that the values and indices get sorted.
+    C.fillComplete(CDomainMap, CRangeMap, params);
```

Comment:
@mhoemmen There's no other reason than that's how it was before :) The other add() just returns a new matrix; that's the only interface that's actually necessary.

Reply:
@brian-kelley cool, sorry, not your fault :-) There may have been an intent at one point for add to have the ability to reuse existing storage in the output matrix.  That might explain why it was an output argument instead of a return value.